README.md
disque-rb
=========

Client for Disque, an in-memory, distributed job queue.

Usage
-----

Create a new Disque client by passing a list of nodes:

```ruby
client = Disque.new(["127.0.0.1:7711", "127.0.0.1:7712", "127.0.0.1:7713"])
```

Alternatively, you can pass a single string with comma-separated nodes:

```ruby
client = Disque.new("127.0.0.1:7711,127.0.0.1:7712,127.0.0.1:7713")
```

Using a single string is useful if you are receiving the list of nodes
from an environment variable.

If the nodes are password protected, you can pass the `AUTH` string:

```ruby
client = Disque.new("127.0.0.1:7711", auth: "e727d1464a...")
```

The client keeps track of which nodes are providing more jobs, and after
a given number operations it tries to connect to the preferred node. The
number of operations for each cycle defaults to 1000, but it can be
configured:

```ruby
client = Disque.new("127.0.0.1:7711", cycle: 20000)
```

Now you can add jobs:

```ruby
client.push("foo", "bar", 100)
```

It will push the job `"bar"` to the queue `"foo"` with a timeout
of 100 ms, and return the id of the job if it was received and
replicated in time.

Disque's `ADDJOB` signature is as follows:

```
ADDJOB queue_name job <ms-timeout>
  [REPLICATE <count>]
  [DELAY <sec>]
  [RETRY <sec>]
  [TTL <sec>]
  [MAXLEN <count>]
  [ASYNC]
```

You can pass any optional arguments as a hash, for example:

```ruby
disque.push("foo", "myjob", 1000, ttl: 1, async: true)
```

Note that `async` is a special case because it's just a flag. That's
why `true` must be passed as its value.

Then, your workers will do something like this:

```ruby
loop do
  client.fetch(from: ["foo"]) do |job|
    # Do something with `job`
  end
end
```

The `fetch` command receives an array of queues, and optionally a
`timeout` (in milliseconds) and the `count` of jobs to retrieve:

```ruby
client.fetch(from: ["bar", "baz"], count: 10, timeout: 2000)
```

Installation
------------

You can install it using rubygems.

```
$ gem install disque
```


CHANGELOG
0.0.6

- Limit exploration to initial node list

0.0.5

- Change API to receive a string of nodes

0.0.4

- Add AUTH option.


tests/disque_test.rb
require_relative "../lib/disque"
require "stringio"
require "fileutils"

module Silencer
  @output = nil

  def self.start
    $olderr = $stderr
    $stderr = StringIO.new
  end

  def self.stop
    @output = $stderr.string
    $stderr = $olderr
  end

  def self.output
    @output
  end
end

DISQUE_NODES = [
  "127.0.0.1:7710",
  "127.0.0.1:7711",
  "127.0.0.1:7712",
  "127.0.0.1:7713",
]

DISQUE_BAD_NODES  = DISQUE_NODES[0,1]
DISQUE_GOOD_NODES = DISQUE_NODES[1,3]

test "raise if connection is not possible" do
  Silencer.start
  assert_raise(ArgumentError) do
    c = Disque.new(DISQUE_BAD_NODES, auth: "test")
  end
  Silencer.stop

  assert_equal "#<Errno::ECONNREFUSED: Can't connect to: disque://:test@127.0.0.1:7710>\n", Silencer.output
end

test "retry until a connection is reached" do
  Silencer.start
  c = Disque.new(DISQUE_NODES, auth: "test")
  Silencer.stop

  assert_equal "#<Errno::ECONNREFUSED: Can't connect to: disque://:test@127.0.0.1:7710>\n", Silencer.output
  assert_equal "PONG", c.call("PING")
end

test "lack of jobs" do
  c = Disque.new(DISQUE_GOOD_NODES, auth: "test")
  reached = false

  c.fetch(from: ["foo"], timeout: 1) do |job|
    reached = true
  end

  assert_equal false, reached
end

test "one job" do
  c = Disque.new(DISQUE_GOOD_NODES, auth: "test")

  c.push("foo", "bar", 1000)

  c.fetch(from: ["foo"], count: 10) do |job, queue|
    assert_equal "bar", job
  end
end

test "multiple jobs" do
  c = Disque.new(DISQUE_GOOD_NODES, auth: "test")

  c.push("foo", "bar", 1000)
  c.push("foo", "baz", 1000)

  jobs = ["baz", "bar"]

  c.fetch(from: ["foo"], count: 10) do |job, queue|
    assert_equal jobs.pop, job
    assert_equal "foo", queue
  end

  assert jobs.empty?
end

test "multiple queues" do
  c = Disque.new(DISQUE_GOOD_NODES, auth: "test")

  c.push("foo", "bar", 1000)
  c.push("qux", "baz", 1000)

  queues = ["qux", "foo"]
  jobs = ["baz", "bar"]

  result = c.fetch(from: ["foo", "qux"], count: 10) do |job, queue|
    assert_equal jobs.pop, job
    assert_equal queues.pop, queue
  end

  assert jobs.empty?
  assert queues.empty?
end

test "add jobs with other parameters" do
  c = Disque.new(DISQUE_GOOD_NODES, auth: "test")

  c.push("foo", "bar", 1000, async: true, ttl: 1)

  sleep 2

  queues = ["foo"]
  jobs = ["bar"]

  result = c.fetch(from: ["foo"], count: 10, timeout: 1) do |job, queue|
    assert_equal jobs.pop, job
    assert_equal queues.pop, queue
  end

  assert_equal ["bar"], jobs
  assert_equal ["foo"], queues
end

test "connect to the best node" do
  c1 = Disque.new([DISQUE_GOOD_NODES[1], DISQUE_GOOD_NODES[0]], cycle: 2, auth: "test")
  c2 = Disque.new([DISQUE_GOOD_NODES[1]], cycle: 2, auth: "test")

  assert c1.prefix != c2.prefix

  # Tamper stats to trigger a reconnection
  c1.stats[c2.prefix] = 10

  c1.push("q1", "j1", 1000)
  c1.push("q1", "j2", 1000)

  c2.push("q1", "j3", 1000)

  c1.fetch(from: ["q1"])
  c1.fetch(from: ["q1"])
  c1.fetch(from: ["q1"])

  # Client should have reconnected
  assert c1.prefix == c2.prefix
end

test "connect to the best node, part 2" do
  c1 = Disque.new([DISQUE_GOOD_NODES[1], DISQUE_GOOD_NODES[0]], cycle: 2, auth: "test")
  c2 = Disque.new([DISQUE_GOOD_NODES[0], DISQUE_GOOD_NODES[1]], cycle: 2, auth: "test")

  assert c1.prefix != c2.prefix

  c1.push("q1", "j1", 0)
  c1.push("q1", "j2", 0)
  c1.push("q1", "j3", 0)

  c2.fetch(from: ["q1"])
  c2.fetch(from: ["q1"])
  c2.fetch(from: ["q1"])

  # Client should have reconnected
  assert c1.prefix == c2.prefix
end

test "recover after node disconnection" do
  Silencer.start

  c1 = Disque.new([DISQUE_GOOD_NODES[0], DISQUE_BAD_NODES[0]], cycle: 2, auth: "test")

  Silencer.stop

  prefix = c1.prefix

  # Tamper stats to trigger a reconnection to a bad node
  c1.stats["fake"] = 10
  c1.nodes["fake"] = DISQUE_BAD_NODES[0]

  # Delete the other nodes just in case
  c1.nodes.delete_if do |key, val|
    key != prefix &&
    key != "fake"
  end

  Silencer.start

  c1.push("q1", "j1", 1000)
  c1.push("q1", "j2", 1000)
  c1.push("q1", "j3", 1000)

  c1.fetch(from: ["q1"])
  c1.fetch(from: ["q1"])
  c1.fetch(from: ["q1"])

  Silencer.stop

  # Prefix should stay the same
  assert_equal prefix, c1.prefix
  assert_equal "#<Errno::ECONNREFUSED: Can't connect to: disque://:test@127.0.0.1:7710>\n", Silencer.output
end

test "federation" do
  c1 = Disque.new([DISQUE_GOOD_NODES[0]], cycle: 2, auth: "test")
  c2 = Disque.new([DISQUE_GOOD_NODES[1]], cycle: 2, auth: "test")

  c1.push("q1", "j1", 0)

  c2.fetch(from: ["q1"], count: 10) do |job, queue|
    assert_equal "j1", job
  end
end

test "ack jobs when block is given" do
  c = Disque.new(DISQUE_GOOD_NODES, auth: "test")

  c.push("q1", "j1", 1000)

  id = nil

  _, id, _ = c.fetch(from: ["q1"]) { |*a| }[0]

  assert id

  info = Hash[*c.call("SHOW", id)]

  if info.any?

    # If the test runs too fast, we may get the job
    # with the status set to "acked"
    assert_equal "acked", info.fetch("state")
  end
end

test "don't ack jobs when no block is given" do
  c = Disque.new(DISQUE_GOOD_NODES, auth: "test")

  c.push("q1", "j1", 1000)

  _, id, _ = c.fetch(from: ["q1"])[0]

  assert id

  info = Hash[*c.call("SHOW", id)]

  assert_equal info.fetch("state"), "active"
end

test "receive a string of comma-separated nodes" do
  nodes = DISQUE_GOOD_NODES.join(",")

  c = Disque.new(nodes, auth: "test")

  assert_equal "PONG", c.call("PING")
  assert_equal DISQUE_GOOD_NODES.size, c.nodes.size
end


lib/disque.rb
require "redic"

class Disque
  ECONN = [
    Errno::ECONNREFUSED,
    Errno::EINVAL,
  ]

  attr :stats
  attr :nodes
  attr :prefix

  # Create a new Disque client by passing a list of nodes.
  #
  #   Disque.new(["127.0.0.1:7711", "127.0.0.1:7712", "127.0.0.1:7713"])
  #
  # Alternatively, you can pass a single string with a comma-separated
  # list of nodes:
  #
  #   Disque.new("127.0.0.1:7711,127.0.0.1:7712,127.0.0.1:7713")
  #
  # For each operation, a counter is updated to signal which node was
  # the originator of the message. Based on that information, after
  # a full cycle (1000 operations, but configurable on initialization)
  # the stats are checked to see what is the most convenient node
  # to connect to in order to avoid extra jumps.
  #
  # TODO Account for timeout
  def initialize(hosts, auth: nil, cycle: 1000)

    # Split a string of hosts if necessary
    if String === hosts
      hosts = hosts.split(",")
    end

    # Cluster password
    @auth = auth

    # Cycle length
    @cycle = cycle

    # Operations counter
    @count = 0

    @hosts = hosts

    # Known nodes
    @nodes = Hash.new

    # Connection stats
    @stats = Hash.new(0)

    # Main client
    @client = Redic.new

    # Scout client
    @scout = Redic.new

    # Preferred client prefix
    @prefix = nil

    explore!
  end

  def url(host)
    if @auth
      sprintf("disque://:%s@%s", @auth, host)
    else
      sprintf("disque://%s", host)
    end
  end

  # Collect the list of nodes and keep a connection to the
  # node that provided that information.
  def explore!

    # Reset nodes
    @nodes.clear

    @hosts.each do |host|
      begin
        @scout.configure(url(host))

        result = @scout.call!("HELLO")

        # For keeping track of nodes and stats, we use only the
        # first eight characters of the node_id. That's because
        # those eight characters are part of the job_ids, and
        # our stats are based on that.
        @prefix = result[1][0,8]

        # Populate cache
        @nodes[@prefix] = host

        # Connect the main client to the last scouted node
        @client.configure(@scout.url)

        @scout.quit

      rescue *ECONN
        $stderr.puts($!.inspect)
      end
    end

    if @nodes.empty?
      raise ArgumentError, "nodes unavailable"
    end
  end

  def pick_client!
    if @count == @cycle
      @count = 0
      prefix, _ = @stats.max { |a, b| a[1] <=> b[1] }

      if prefix != @prefix
        host = @nodes[prefix]

        if host

          # Reconfigure main client
          @client.configure(url(host))

          # Save current node prefix
          @prefix = prefix

          # Reset stats for this new connection
          @stats.clear
        end
      end
    end
  end

  # Run commands on the active connection. If the
  # connection is lost, new connections are tried
  # until all nodes become unavailable.
  def call(*args)
    @client.call!(*args)
  rescue *ECONN
    explore!
    retry
  end

  # Disque's ADDJOB signature is as follows:
  #
  #     ADDJOB queue_name job <ms-timeout>
  #       [REPLICATE <count>]
  #       [DELAY <sec>]
  #       [RETRY <sec>]
  #       [TTL <sec>]
  #       [MAXLEN <count>]
  #       [ASYNC]
  #
  # You can pass any optional arguments as a hash,
  # for example:
  #
  #     disque.push("foo", "myjob", 1000, ttl: 1, async: true)
  #
  # Note that `async` is a special case because it's just a
  # flag. That's why `true` must be passed as its value.
  def push(queue_name, job, ms_timeout, options = {})
    command = ["ADDJOB", queue_name, job, ms_timeout]
    command += options_to_arguments(options)

    call(*command)
  end

  def fetch(from: [], count: 1, timeout: 0)
    pick_client!

    jobs = call(
      "GETJOB",
        "TIMEOUT", timeout,
        "COUNT", count,
        "FROM", *from)

    if jobs then
      @count += 1

      jobs.each do |queue, msgid, job|

        # Update stats
        @stats[msgid[2,8]] += 1

        if block_given?

          # Process job
          yield(job, queue)

          # Remove job
          call("ACKJOB", msgid)
        end
      end
    end

    return jobs
  end

  def options_to_arguments(options)
    arguments = []

    options.each do |key, value|
      if value == true
        arguments.push(key)
      else
        arguments.push(key, value)
      end
    end

    return arguments
  end

  def quit
    @client.quit
  end
end
